### DISCUSSION TITLE
Issue with intrinsic "to_v32bfloat16"

### DISCUSSION THREAD

**Discussion by giacomo-brunetta** [AMDResearch/Riallto#64] [Q&A] [answered]
Hello everyone,
I would like to report an issue that I found while working with bfloat16s on the AIE-ML.
The function `v32bfloat16 to_v32bfloat16 | ( | v32accfloat | acc | )`  that is declared in the AI Engine-ML Intrinsics User Guide  (v2023.2) is not recognized by the compiler.

I figured it out writing this code, that compiles only for N = 16

```
template<int N=16>
inline aie::vector<bfloat16, N> int16_to_bfloat16(aie::vector<int16, N> x){
    aie::vector<float,N> x_float = aie::to_float(x);
    if constexpr (N == 16){
        aie::accum<accfloat,N> x_acc;
        x_acc = v16accfloat(x_float);
        return to_v16bfloat16(x_acc);
    }
    else{ //N == 32
        aie::accum<accfloat,N> x_acc;
        x_acc = v32accfloat(x_float);
        return to_v32bfloat16(x_acc);
    }
}
```

The error  that is thrown is the following

```
logv.cc:50:16: error: use of undeclared identifier 'to_v32bfloat16'
        return to_v32bfloat16(x_acc);
```

**Comment by giacomo-brunetta** [comment]
I found a dirty walkaround by writing it as
```
inline aie::vector<bfloat16, 32> to_v32bfloat16(aie::vector<float,32> x){
    aie::vector<int16, 64> temp = v64int16(x);
    aie::vector<int16, 32> temp_trunc = aie::filter_odd(temp);
    return v32bfloat16(temp_trunc);
}
```

But the truncation always "floors" the mantissa, rather than approximating it correctly

**Reply by mariodruiz** [reply, marked as answer]
In Riallto we use the tools 2023.1, so this could be the issue.

**Reply by giacomo-brunetta** [reply]
Ok, this explains why it is not working. I was using the 2023.2 because it is the closest to the 2023.1 available here https://www.xilinx.com/htmldocs/aiengine_intrinsics_start.html
